packages/edit-site/src/components/pattern-actions/index.js

```diff
@@ -0,0 +1,281 @@
+/**
+ * External dependencies
+ */
+import { paramCase as kebabCase } from 'change-case';
+
+/**
+ * WordPress dependencies
+ */
+import { useDispatch, useSelect } from '@wordpress/data';
+import { store as coreStore } from '@wordpress/core-data';
+import { useState } from '@wordpress/element';
+import { __, sprintf } from '@wordpress/i18n';
+import {
+	DropdownMenu,
+	MenuGroup,
+	MenuItem,
+	__experimentalConfirmDialog as ConfirmDialog,
+} from '@wordpress/components';
+import { moreVertical } from '@wordpress/icons';
+import { store as noticesStore } from '@wordpress/notices';
+import { decodeEntities } from '@wordpress/html-entities';
+import { store as reusableBlocksStore } from '@wordpress/reusable-blocks';
+import { downloadBlob } from '@wordpress/blob';
+import { getQueryArgs } from '@wordpress/url';
+
+/**
+ * Internal dependencies
+ */
+import { store as editSiteStore } from '../../store';
+import isTemplateRemovable from '../../utils/is-template-removable';
+import isTemplateRevertable from '../../utils/is-template-revertable';
+import RenameMenuItem from './rename-menu-item';
+import DuplicateMenuItem from './duplicate-menu-item';
+import {
+	TEMPLATE_POST_TYPE,
+	TEMPLATE_PART_POST_TYPE,
+	POST_TYPE_LABELS,
+	PATTERN_TYPES,
+} from '../../utils/constants';
+
+export default function PatternActions( {
+	item,
+	className,
+	toggleProps = {},
+	popoverProps = {},
+	onRemove,
+	icon = moreVertical,
+} ) {
+	const record = useSelect(
+		( select ) => {
+			if ( item?.type && item?.id ) {
+				return select( coreStore ).getEntityRecord(
+					'postType',
+					item?.type,
+					item?.id
+				);
+			}
+			return null;
+		},
+		[ item?.type, item?.id ]
+	);
+	const { removeTemplate, revertTemplate } = useDispatch( editSiteStore );
+	const { saveEditedEntityRecord } = useDispatch( coreStore );
+	const { createSuccessNotice, createErrorNotice } =
+		useDispatch( noticesStore );
+	const { __experimentalDeleteReusableBlock } =
+		useDispatch( reusableBlocksStore );
+	// Only custom patterns or custom template parts can be renamed or deleted.
+	const isUserPattern = record?.type === PATTERN_TYPES.user;
+	const isNonUserPattern = item?.type === PATTERN_TYPES.theme;
+	const isTemplate = record?.type === TEMPLATE_POST_TYPE;
+	const isTemplatePart = record?.type === TEMPLATE_PART_POST_TYPE;
+
+	if (
+		! isTemplatePart &&
+		! isTemplate &&
+		! isUserPattern &&
+		! isNonUserPattern
+	) {
+		return null;
+	}
+
+	const isRemovable = isTemplateRemovable( record );
+	const isRevertable = isTemplateRevertable( record );
+	const isEditable = isUserPattern || isRemovable;
+	// Only patterns and template parts can be duplicated for now.
+	const isDuplicable = isUserPattern || isNonUserPattern || isTemplatePart;
```

Comment:
With this condition, it seems that duplication of template parts of the theme is also allowed. Is this intentional?

![duplicate-template-part](https://github.com/WordPress/gutenberg/assets/54422211/232439b1-48fd-45b1-b779-f99b0cb36559)

Furthermore, hybrid themes seem to allow duplication of theme template parts. You can check it by following the steps below.

- Access `http://localhost:8889/wp-admin` (instead of `http://localhost:8888/wp-admin`)
- Activate Emptyhybrid theme (this theme supports block-template-parts)
- Access Appearance > Template Parts
  - Expected (trunk): No action button is displayed.
  - Actual (this PR): Can duplicate template parts of the theme

Reply:
> With this condition, it seems that duplication of template parts of the theme is also allowed. Is this intentional?

Yes. It looks like we can duplicate theme template parts on trunk as well so I wanted to harmonize the behaviour.

<img width="200" alt="Screenshot 2023-12-11 at 6 56 57 pm" src="https://github.com/WordPress/gutenberg/assets/6458278/b26cfc70-3dc8-4a89-8e7b-62a7e5787a4b">

Reply:
> It looks like we can duplicate theme template parts on trunk as well so I wanted to harmonize the behaviour.

I had missed being able to do something like that 😅 Then I think this behavior is correct. However, I think we need to disable the duplicate action for classic themes that support block template parts.

Reply:
Good call. I've prevented non-block based themes from rendering the drop down. Is that the best way forward do you think?